cmsplugin_feed_ai/feed.py

```diff
@@ -0,0 +1,171 @@
+from twython import Twython
+from warnings import warn
+
+
+DEFAULT_CACHE_TIME = 3600  # 1 hour
+
+
+class FeedDisplayItem(object):
+    created_at = None
+    html_content = ""
+    content_link = ""
+    profile_name = ""
+    source = ""
+    source_id = ""
+    text_content = ""
+
+    def __init__(self, feed_obj, profile_name, source):
+        """
+        Initializer display item from raw feed object provided by
+        service specific API tool.
+        """
+        self.profile_name = profile_name
+        self.source = source
+        if source == "twitter":
+            self._init_from_twitter_object(feed_obj)
+        elif source == "facebook":
+            self._init_from_facebook_object(feed_obj)
+
+    def __str__(self):
+        return "%s %s" % (self.source, self.text_content[:20])
+
+    def __repr__(self):
+        return "<FeedDisplayItem {'source': %s, 'text_content': %s, 'created_at': %s}>" % (
+            self.source, self.text_content, self.created_at
+        )
+
+    def _init_from_twitter_object(self, obj):
+        self.created_at = parse_datetime(obj.get("created_at", ""))
+        self.html_content = Twython.html_for_tweet(obj)
+        self.source_id = obj.get("id_str", "")
+        self.text_content = obj.get("text", "")
+
+    def _init_from_facebook_object(self, obj):
+        self.created_at = parse_datetime(obj.get("created_time", ""))
+        self.content_link = obj.get("link", "")
+        self.source_id = obj.get("id", "")
+        self.text_content = obj.get(
+            "message",
+            obj.get("status_type", "").replace("_", " ").capitalize
+        )
+
+    @property
+    def service_url(self):
+        """
+        URL for the source service
+        """
+        return "https://%s.com" % self.source
+
+    @property
+    def profile_url(self):
+        """
+        URL which points to the profile related to this feed item.
+        """
+        return "%s/%s" % (self.service_url, self.profile_name)
+
+    @property
+    def permalink(self):
+        """
+        Permalink which points to the original feed object in external service.
+        """
+        if self.source == "twitter":
+            return "%s/status/%s" % (self.profile_url, self.source_id)
+        elif self.source == "facebook":
+            return "%s/posts/%s" % (
+                self.profile_url, self.source_id.split("_")[-1]
+            )
+
+    @property
+    def link(self):
+        return self.content_link or self.permalink
+
+
+def get_feed_settings(key=None, default=None):
+    feed_settings = getattr(settings, "AI_FEED_SETTINGS", {})
+    if key not in feed_settings:
+        warn("Missing %r settings in AI_FEED_SETTINGS" % key)
+        return default
+    return feed_settings.get(key, default)
+
+
+def get_feed_cache_key(service):
+    """Return a unique cache key for given service"""
+    return "cmsplugin_feed_ai_%s_feed" % service
+
+
+def _fetch_twitter_profile_feed(profile_name, limit):
```

Comment:
Twitter-specific code should live in one submodule (`cmsplugin_feed_ai.providers.twitter`?) and Facebook-specific in another, I think.